swarm_dial.go

```diff
@@ -345,69 +348,70 @@ func (s *Swarm) filterKnownUndialables(addrs []ma.Multiaddr) []ma.Multiaddr {
 
 func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.Multiaddr) (transport.Conn, error) {
 	log.Debugf("%s swarm dialing %s", s.local, p)
-
 	ctx, cancel := context.WithCancel(ctx)
-	defer cancel() // cancel work when we exit func
-
-	// use a single response type instead of errs and conns, reduces complexity *a ton*
-	respch := make(chan dialResult)
-
-	defaultDialFail := fmt.Errorf("failed to dial %s (default failure)", p)
-	exitErr := defaultDialFail
-
+	defer cancel()
 	defer s.limiter.clearAllPeerDials(p)
-
-	var active int
-	for remoteAddrs != nil || active > 0 {
-		// Check for context cancellations and/or responses first.
+	var (
+		dialResults = make(chan dialResult)
+		lastDialErr error
+		activeDials int
+	)
+	onDialResult := func(dr dialResult) (retConn bool) {
+		activeDials--
+		if dr.Err != nil {
+			log.Infof("got error on dial to %s: %s", dr.Addr, dr.Err)
+			// Errors are normal, lots of dials will fail. Keep one to return if
+			// no dials succeed.
+			lastDialErr = dr.Err
+			return false
+		} else {
+			return true
+		}
+	}
+	ctxDoneRetErr := func() error {
+		if lastDialErr != nil {
+			return lastDialErr
+		} else {
+			return ctx.Err()
+		}
+	}
+	for remoteAddrs != nil || activeDials > 0 {
+		// Try to proceed on context completion or a dial result before dialing
+		// more addresses.
 		select {
 		case <-ctx.Done():
-			if exitErr == defaultDialFail {
-				exitErr = ctx.Err()
-			}
-			return nil, exitErr
-		case resp := <-respch:
-			active--
-			if resp.Err != nil {
-				log.Infof("got error on dial to %s: %s", resp.Addr, resp.Err)
-				// Errors are normal, lots of dials will fail
-				exitErr = resp.Err
-			} else if resp.Conn != nil {
-				return resp.Conn, nil
+			return nil, ctxDoneRetErr()
+		case dr := <-dialResults:
+			if onDialResult(dr) {
+				return dr.Conn, nil
+			} else {
```

Comment:
This `else` is superfluous.

Reply:
Because this is a pattern match, and not a guard, like you would normally apply this rule in Go. It's a stylistic choice I've borrowed from Haskell. Neither boolean value is exceptional here, I'm only acting on the control flow indication provided.

Reply:
I removed it anyway, it looks nicer.